ts/private/ts_project_worker.js

```diff
@@ -540,23 +538,53 @@ function createProgram(args, initialInputs) {
     }
 }
 
+// How much (%) of memory should be free at all times. 
+const NEAR_OOM_ZONE = 20
+
+function isNearOomZone() {
+    const stat = v8.getHeapStatistics();
+    const used = (100 / stat.heap_size_limit) * stat.used_heap_size
+    return 100 - used < NEAR_OOM_ZONE
+}
+
+/** @type {Map<string, ReturnType<createProgram> & {previousInputs?: import("@bazel/worker").Inputs}>} */
+const workers = new Map();
+
+function sweepLeastRecentlyUsedWorkers() {
+    for (const [k, w] of workers) {
+        w.program.close();
+        workers.delete(k);
+        // stop killing workers as soon as the worker is out the oom zone 
+        if (!isNearOomZone()) {
+            break;
+        }
+    }
+}
 
 function getOrCreateWorker(args, inputs) {
     const project = args[args.indexOf('--project') + 1]
     const outDir = args[args.lastIndexOf("--outDir") + 1]
     const declarationDir = args[args.lastIndexOf("--declarationDir") + 1]
     const rootDir = args[args.lastIndexOf("--rootDir") + 1]
     const key = `${project} @ ${outDir} @ ${declarationDir} @ ${rootDir}`
-    if (!workers.has(key)) {
+
+    let worker = workers.get(key)
+    if (!worker) {
+        if (isNearOomZone()) {
+            sweepLeastRecentlyUsedWorkers();
+        }
         const { program, host, checkAndApplyArgs } = createProgram(args, inputs);
         host.debuglog(`Created a new worker for ${key}`);
-        workers.set(key, {
+        worker = {
             program,
             host,
             checkAndApplyArgs
-        });
+        }
+    } else {
+        workers.delete(key)
```

Comment:
When would it be in the map but falsy?

Reply:
Oh... this is to ensure it gets put at the end again?

Reply:
yeap.